smalltalk.asciidoc

```diff
@@ -0,0 +1,1617 @@
+[source, pony]
+----
+  be add_order(order: Order) =>
+    if (not _sym.eq(order.sym)) then return end
+    try
+      _map.insert(order.order_id, order)
+      _views_do(lambda (v: View tag) (order) => v.add_order(order) end)
+    else
+      _trace("error inserting order")
+    end
+----
+
+Updating is somewhat more interesting. `HashMap.update()` returns either the value previously stored for a key or `None` (see the standard library's http://www.ponylang.org/ponyc/collections-HashMap/#public-functions[documentation]). Therefore we use `match` to check what is returned. Match expressions can be seen as an advanced form of `switch` statements in other languages. In Pony, we can match against types, values, tuples--but we'll get to that later.  
+[source, pony]
+----
+  be edit_order(order: Order) =>
+    if (not _sym.eq(order.sym)) then return end    
+    match _map.update(order.order_id, order)
+    | let old: Order =>
+      _views_do(lambda (v: View tag) (old, order) => v.update_order(old, order) end)
+    else
+      _trace("error updating order")
+    end
+----
+
+The `remove` operation on the underlying map also returns an interesting value: a tuple composed of the key and the value of the removed item, in that order. We're interested in the value, which we extract it from the tuple with `._2`.
+[source, pony]
+----
+  be delete_order(order: Order) =>
+    if (not _sym.eq(order.sym)) then return end    
+    try
+      let removed = _map.remove(order.order_id)._2
+      _views_do(lambda (v: View tag)(removed) => v.delete_order(removed) end)
+    else
+      _trace("error deleting order")
+    end
+----
+
+Lastly, just to demonstrate the use of lambdas, we have factored out applying functions to the collection of views. This function is private to the book.
+[source, pony]
+----
+  fun _views_do(f: {(View tag)}) =>
+    """
+    Applies function f to all views.
+    """
+    for each in _views.values() do f(each) end
+----
+
+Noticed something missing from the book actor? No explicit threading or message queuing code. It's all done behind the scenes by Pony.
+
+=== Records
+We declare a new type alias for our two kinds of records.
+[source, pony]
+----
+type Record is (ByPriceRecord | ByClientRecord)
+----
+
+The pattern of annotating these classes with `val` has been applied here for the same reasons mentioned earlier. The `string()` method is used to pretty-print record instances when debugging.
+
+`string()` returns a string that is currently unassigned (nothing points to it), denoted by `^`. It can be assigned only to an isolated variable, an `iso`. How is the `String ref` constructed in the method converted into a `String iso^`? The string is created and modified inside a `recover` expression. Only within `recover` is the string mutable, outside it is _frozen_, seen as having a `val` rcap.
```

Comment:
"Only within `recover` is the string mutable, outside it is _frozen_, seen as having a `val` rcap."

Once the `String` leaves the `recover` block it has a reference capability of `iso^`, not `val`. Maybe I'm missing what you mean when you use the word "seen" here.